plasmapy_nei/eigen/eigenclass.py

```diff
@@ -0,0 +1,360 @@
+        )
+        self._eigenvalues = np.ndarray(
+            shape=(self.ntemp, self.nstates), dtype=np.float64
+        )
+
+        self._eigenvectors = np.ndarray(
+            shape=(self.ntemp, self.nstates, self.nstates), dtype=np.float64,
+        )
+
+        self._eigenvector_inverses = np.ndarray(
+            shape=(self.ntemp, self.nstates, self.nstates), dtype=np.float64,
+        )
+
+        self._ionization_rate[:, :] = c_rate[:, :]
+        self._recombination_rate[:, :] = r_rate[:, :]
+
+        # Define the coefficients matrix A. The first dimension is
+        # for elements, and the second number of equations.
+
+        nequations = self.nstates
+        coefficients_matrix = np.zeros(
+            shape=(self.nstates, nequations), dtype=np.float64
+        )
+
+        # Enter temperature loop over the whole temperature grid
+
+        for temperature_index in range(self.ntemp):
+
+            # Ionization and recombination rate at Te(temperature_index)
+            carr = c_rate[temperature_index, :]
+            rarr = r_rate[temperature_index, :]
+
+            eqi = _get_equilibrium_charge_states(
+                carr, rarr, self.element.atomic_number,
+            )
+
+            for ion in range(1, self.nstates - 1):
+                coefficients_matrix[ion, ion - 1] = carr[ion - 1]
+                coefficients_matrix[ion, ion] = -(carr[ion] + rarr[ion])
+                coefficients_matrix[ion, ion + 1] = rarr[ion + 1]
+
+            # The first row
+            coefficients_matrix[0, 0] = -carr[0]
+            coefficients_matrix[0, 1] = rarr[1]
+
+            # The last row
+            coefficients_matrix[self.nstates - 1, self.nstates - 2] = carr[
+                self.nstates - 2
+            ]
+            coefficients_matrix[self.nstates - 1, self.nstates - 1] = -rarr[
+                self.nstates - 1
+            ]
+
+            # Compute eigenvalues and eigenvectors
+            eigenvalues, eigenvectors = LA.eig(coefficients_matrix)
+
+            # Rearrange the eigenvalues.
+            idx = np.argsort(eigenvalues)
+            eigenvalues = eigenvalues[idx]
+            eigenvectors = eigenvectors[:, idx]
+
+            # Compute inverse of eigenvectors
+            v_inverse = LA.inv(eigenvectors)
+
+            # transpose the order to as same as the Fortran Version
+            eigenvectors = eigenvectors.transpose()
+            v_inverse = v_inverse.transpose()
+
+            # Save eigenvalues and eigenvectors into arrays
+            for j in range(self.nstates):
+                self._eigenvalues[temperature_index, j] = eigenvalues[j]
+                self._equilibrium_states[temperature_index, j] = eqi[j]
+                for i in range(self.nstates):
+                    self._eigenvectors[temperature_index, i, j] = eigenvectors[i, j]
+                    self._eigenvector_inverses[temperature_index, i, j] = v_inverse[
+                        i, j
+                    ]
```

Comment:
I haven't looked at whether we'd have any performance issues here, but this code looks like a good candidate for "extract function and throw numba at it". Few self dependencies, and those that are there are integers. Could get some pretty healthy speedups here if you need them!

Reply:
I generally agree, but I don't think we'd see much speedup since the for-loops will be small.  If I remember correctly, `nstates < 30`. Not sure about `ntemp` though.